app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,264 @@
+        }
+
+        fun processMessage(msg: Message) {
+            if (msg.what == MSG_ADD_ACTIVE_NETWORK) {
+                processActiveNetwork()
+            } else if (msg.what == MSG_ADD_ALL_NETWORKS) {
+                processAllNetwork()
+            }
+        }
+
+        /**
+         * Handles the active network.
+         * Get active network from connectivity-manager
+         * Set the vpn underlying network when the current active network is different from
+         * already assigned active network or if there is difference in the available network set
+         */
+        private fun processActiveNetwork(){
+            val activeNetwork: Network? = connectionMonitor.connectivityManager.activeNetwork
+            val isNewNetwork = connectionMonitor.currentActiveNetwork?.networkHandle != activeNetwork?.networkHandle
+            connectionMonitor.currentActiveNetwork = activeNetwork
+            if(!isNewNetwork){
+                return
+            }
+            if (activeNetwork != null) {
+                if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - process message ${connectionMonitor.connectivityManager.getNetworkInfo(activeNetwork)?.typeName.toString()}")
+                connectionMonitor.networkListener.onUpdateActiveNetwork(activeNetwork)
+            } else {
+                connectionMonitor.networkListener.onNetworkDisconnected()
+            }
+            if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - ${connectionMonitor.connectivityManager.getNetworkInfo(activeNetwork)?.typeName.toString()}")
+            if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - ${connectionMonitor.connectivityManager.getNetworkInfo(connectionMonitor.currentActiveNetwork)?.typeName.toString()}")
+        }
+
+
+
+        /**
+         * Adds all the available network to the underlying network.
+         */
+        private fun processAllNetwork() {
+            val newActiveNetwork = connectionMonitor.connectivityManager.activeNetwork
+            var isNewNetwork = connectionMonitor.currentActiveNetwork != newActiveNetwork
+            connectionMonitor.currentActiveNetwork  = newActiveNetwork
+            val addedNetworkList = createNetworksSet(newActiveNetwork)
+
+            isNewNetwork = isNewNetwork || hasDifference(connectionMonitor.currentNetworkList, addedNetworkList)
+            if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - process message MESSAGE_AVAILABLE_NETWORK")
+
+            if(!isNewNetwork) return
+
+            if (connectionMonitor.currentNetworkList.size > 0 && addedNetworkList.size > 0) {
+                if (DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - Current Network -" +
+                 "${connectionMonitor.connectivityManager.getNetworkInfo(connectionMonitor.currentNetworkList.elementAt(0))?.typeName.toString()}, " +
+                  "already added network- ${connectionMonitor.connectivityManager.getNetworkInfo(addedNetworkList.elementAt(0))?.typeName.toString()}")
+            }
```

Comment:
not required. add to the log statement above. I believe, Kotlin, by default, unwraps contents of a Set for logs.

Reply:
Kotlin unwraps and printing the id's in the set and also the network object has log of property. To know whether the network type, we need to get the networkInfo's typename property.